hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/DFSInputStream.java

```diff
@@ -1494,19 +1533,39 @@ private int pread(long position, ByteBuffer buffer)
       try {
         if (dfsClient.isHedgedReadsEnabled() && !blk.isStriped()) {
           hedgedFetchBlockByteRange(blk, targetStart,
-              targetEnd, buffer, corruptedBlocks);
+              targetEnd, buffer, corruptedBlocks, ioExceptionMap);
         } else {
           fetchBlockByteRange(blk, targetStart, targetEnd,
-              buffer, corruptedBlocks);
+              buffer, corruptedBlocks, ioExceptionMap);
         }
-      } finally {
+      } catch (IOException e) {
+        // When we reach here, it means we fail to fetch the current block from all available
+        // datanodes. Send IOExceptions in ioExceptionMap to the log and rethrow the exception to
+        // fail this request.
+        String msg = String.format("Failed to read from all available datanodes for block %s "
```

Comment:
This error handling is duplicated in multiple places. Can we abstract it in a common function?

Reply:
Thanks for the suggestion! Fixed.